src/components/Chart/data/dataAccumulator.js

```diff
@@ -33,46 +33,48 @@
  * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
-/* eslint no-bitwise: off */
 /* eslint no-plusplus: off */
 
+import bitDataAccumulator from './bitDataAccumulator';
+import noOpBitDataProcessor from './noOpBitDataProcessor';
+
 import { options, timestampToIndex, nbDigitalChannels } from '../../../globals';
-import { doubleBitValue } from '../../../utils/bitConversion';
 
 const emptyArray = () =>
     [...Array(4000)].map(() => ({ x: undefined, y: undefined }));
 
-const digitalOnValue = 0.4;
-const digitalOffValue = -digitalOnValue;
-
 export default () => ({
-    lineData: emptyArray(),
-    bitsData: [...Array(nbDigitalChannels)].map(() => emptyArray()),
-    bitIndexes: new Array(nbDigitalChannels),
-    bitAccumulator: new Array(nbDigitalChannels),
+    ampereLineData: emptyArray(),
+    bitDataAccumulator: bitDataAccumulator(),
+    noOpBitDataProcessor: noOpBitDataProcessor(),
+    bitStateAccumulator: new Array(nbDigitalChannels),
 
-    process(begin, end, numberOfBits, len, windowDuration) {
-        const { bits, data, index } = options;
+    process(begin, end, digitalChannelsToCompute, len, windowDuration) {
+        const { data, index } = options;
+        const bitDataProcessor =
+            digitalChannelsToCompute.length > 0
+                ? this.bitDataAccumulator
+                : this.noOpBitDataProcessor;
 
         const originalIndexBegin = timestampToIndex(begin, index);
         const originalIndexEnd = timestampToIndex(end, index);
         const step = (originalIndexEnd - originalIndexBegin) / len;
 
         let mappedIndex = 0;
-        this.bitIndexes.fill(0);
+        let timestamp;
+
+        bitDataProcessor.initialise(digitalChannelsToCompute);
 
         for (
             let originalIndex = originalIndexBegin;
             mappedIndex < len + len;
             ++mappedIndex, originalIndex += step
         ) {
-            const timestamp =
-                begin + windowDuration * (mappedIndex / (len + len));
+            timestamp = begin + windowDuration * (mappedIndex / (len + len));
```

Comment:
I wonder why you've lifted up _timestamp_ out of the loop, since it was lifted in both files. Is it to gain a bit of performance? If so, there are many more that could be lifted.

Reply:
That was a residue from a different solution when I used `timestamp` after the loop. It is not needed anymore and I changed it in a4f8412. I do not believe this changes the performance.